### ISSUE TITLE
Reproduce the result of Figure 5(a)

### ISSUE THREAD

**Issue by TianyuanYu** [pathak22/noreward-rl#21]
Hi, 
Thank you for the code.

But when I want to reproduce the result of Figure 5(a) in the paper, I find the results of A3C and ICM+A3C are very similar. 

I use child mode because I am not familiar with tmux.

The only difference between the two algorithms in the code is --unsup in train.py. ICM+A3C is action while A3C is None. Is that right?

Thank you in advance.

**Comment by pathak22** [comment]
Yes, that's right. I am (and few other users I know) have been able to reproduce all the results in paper so far, so it might be some minor issue in your case. Are you using python2? 

You should not need child mode and be able to run the code as it is in github README. You can perhaps debug if the workers inside tmux are using the virutalenv python without needing child mode?

**Comment by TianyuanYu** [comment]
Thanks for your reply.

I am sure I have installed all the dependencies as README. I use python 2 to run the program.

In tmux, the new screen cannot automatically enter the virtual environment, leading to errors.

I make the following changes to fit tensorflow 1.4.0 so that I can use my own environment rather than the provided one (only change versions of dependencies):
In model.py, 
(1)
self.invloss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(logits, aindex), name="invloss") 
-> 
self.invloss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=aindex), name="invloss")
(2) change all 'tf.concat(1,[])' to 'tf.concat([],1)'
(3)rnn.rnn_cell.BasicLSTMCell(size, state_is_tuple=True) ->
rnn.BasicLSTMCell(size, state_is_tuple=True)

state_in = rnn.rnn_cell.LSTMStateTuple(c_in, h_in) ->
rnn.BasicLSTMCell(size, state_is_tuple=True)

What is more, I only use 3 workers to run both a3c and curiosity.

The results are still similar. 

For sparse and very_sparse, I have reproduced the result like you did in the paper [in child mode]. (curiosity is much better than a3c, but not increase like stairs)

Could you analyze which part may be the reason in my case?

Thanks in advance.

**Comment by pathak22** [comment]
3 workers are not enough for a3c to work which forms the backbone algorithm in this case. You would need 16 or more workers.

**Comment by TianyuanYu** [comment]
Thanks. I will try more workers and report the result here.

**Comment by TianyuanYu** [comment]
Hi,

I have tried using 16 workers, only to find the speed has raised more than 5 times. The steps_per_episode is worse than that produced by 3 workers when both converge. That maybe caused by more random initial positions in workers? 

Is that possible that the reason is the version of tensorflow?

Thanks

**Comment by AdamStelmaszczyk** [comment]
> Is that possible that the reason is the version of tensorflow?

I think it is possible. This repo is based on [`universe-starter-agent`](https://github.com/openai/universe-starter-agent), where there are several people reporting worse results:

https://github.com/openai/universe-starter-agent/issues/83
https://github.com/openai/universe-starter-agent/issues/52
https://github.com/openai/universe-starter-agent/issues/86

Quote from the last one:

>  I downgraded my tensorflow from 1.0.0 to 0.11.0 and now it trains successfully as shown in the Readme page.

I also fought with versions for some time, since the visualization looks wrong on my machine (it looks wrong the same way for both `noreward-rl` and `universe-starter-agent`), I described it here: https://github.com/openai/universe-starter-agent/issues/133

Any ideas how to fix this are welcome.

**Comment by TianyuanYu** [comment]
@AdamStelmaszczyk 
Hi,
Thanks for your answer.

Have you tried using child mode in the virtual environment provided by this project? Because as I tried, there is no differences between modes. 

I will do this and report here if you have not. 

Thanks and Happy New Year

**Comment by AdamStelmaszczyk** [comment]
> Have you tried using child mode in the virtual environment provided by this project? Because as I tried, there is no differences between modes.

Yes, I tried. When host has the same packages installed as the `curiosty` virtualenv then I see no difference between tmux and child modes. The visualization looks wrong in the same way.

**Comment by TianyuanYu** [comment]
Hi, 

I tried using child mode, and no more changes in codes.

The result in the paper has been reproduced even if there are only 3 workers.

Thank you for your help. I am going to close this.
